### DISCUSSION TITLE
CP Package Loading (js v.s. css)

### DISCUSSION THREAD

**Discussion by RK311y** [ExpressionEngine/ExpressionEngine#3006] [Feature Requests]
**Description of the problem**

The legacy library [cp](https://github.com/ExpressionEngine/ExpressionEngine/blob/63c0f8cc181e83aa7dbe58c3245d004c54c7bfd3/system/ee/legacy/libraries/Cp.php#L14), located at `system/ee/legacy/libraries/Cp.php`,  exposes two methods for loading add-on package resources, one for JavaScript files and one for CSS files:

- [`load_package_js()`](https://github.com/ExpressionEngine/ExpressionEngine/blob/63c0f8cc181e83aa7dbe58c3245d004c54c7bfd3/system/ee/legacy/libraries/Cp.php#L821)
- [`load_package_css()`](https://github.com/ExpressionEngine/ExpressionEngine/blob/63c0f8cc181e83aa7dbe58c3245d004c54c7bfd3/system/ee/legacy/libraries/Cp.php#L837)

Ideally, both types of resources require the same procedure to render their contents to the output. The difference being that JavaScript files are located in a directory named `./addon/javascript/..` and have the file name extension `.js`, and CSS StyleSheet files are located in a directory `./addon/css/...` and have the file name extension `.css`. 

Both of these file types can be managed in such a way that loading a single file twice does not result in duplicate output of that file. For example, the `load_package_js()` accomplishes this by indexing the file names by the parent addon: 

https://github.com/ExpressionEngine/ExpressionEngine/blob/63c0f8cc181e83aa7dbe58c3245d004c54c7bfd3/system/ee/legacy/libraries/Cp.php#L547-L551

This maintains a unique dataset of JavaScript files to be loaded, discouraging duplicates. The variable definition for this array can be see here:

https://github.com/ExpressionEngine/ExpressionEngine/blob/63c0f8cc181e83aa7dbe58c3245d004c54c7bfd3/system/ee/legacy/libraries/Cp.php#L26-L32

Now looking at the `load_package_css()` method, you will notice that the package resource is automatically appended to the head (of the output). No work is done to prevent duplicate css link elements in the head of the html output. 

https://github.com/ExpressionEngine/ExpressionEngine/blob/63c0f8cc181e83aa7dbe58c3245d004c54c7bfd3/system/ee/legacy/libraries/Cp.php#L829-L849

**How To Reproduce**

```php
// even with two identical calls, duplicate output is avoided 
ee()->cp->load_package_js('main.min');
ee()->cp->load_package_js('main.min');

// two identical calls to the css loader results in duplicate css link elements in the head of the html output

ee()->cp->load_package_css('main.min');
ee()->cp->load_package_css('main.min');
```

The problem arose when loading css packages from within a custom field type. The field type accepts `grid` content, therefore, every row in the grid which had my custom field added an extra css link element to the head of the output html.  

**Environment Details:**
 - Version: Probably all, it’s in the cp legacy library

**Possible Solution**

Implement a similar technique for managing loaded css files like one that is done for JavaScript files. 

**Additional context**

Currently, JavaScript files utilize the combo loader. Allowing the combination of footer scripts to be loaded in a single call. I believe there is an identical combo loader for css files, maybe utilize that.

**Comment by RK311y** [comment]
CSS combo loader (at least I think I found it)


https://github.com/ExpressionEngine/ExpressionEngine/blob/0e068ab68919d8c4ff53047e3c662b2f04886e92/system/ee/ExpressionEngine/Addons/channel/mod.channel.php#L4942-L4970

**Comment by intoeetive** [comment]
@RK311y thank you so much for investigating all of this!

Does this result in some kind of issue for you (performance, wrong styles used etc.) - or you just see this as path for improvement?

**Comment by RK311y** [comment]
@intoeetive the entirety of this would fall under the scope of performance. And I guess you would consider it a bug to have duplicate css outputs, but it does not result in any wrong styles. 

My interpretation on the combo loader is such that maximum performance is achieved due to caching the response of multiple files by a single request.

**Comment by robsonsobral** [comment]
As order means on CSS, this makes me wonder... Is it possible to load repeated CSS files among different ones? Something like:

- `file1.css`;
- `file2.css`;
- `file1.css`.

If so, to fix this can actually bring issues to the styles of the CP. I'm not saying we shouldn't fix it. We must. However, we need to use the CP for sometime after fixing it to make sure nothing moved around or became invisible.
